keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,431 @@
+
+
+### Add recording rules consistent with this metric to describe actual resource usage
+
+In order to make these metrics useful for calculation of actual usage, we would represent the current usage in a form that can easily be queried by applying recording rules in the standard Prometheus stack and equivalents in other systems without having to apply complex calculations.
+
+This section is normative and will be incrementally approached during alpha - not all resource metrics must appear in this format.
+
+* `kube_running_pod_resource_usage` - the resources currently consumed by a running pod
+  * The value would be the unit of consumption for that pod on that node
+  * It should report the following labels to the user:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `node` - the node the resource is scheduled to
+    * `resource` - the name of the resource as it would be described in the pod's `spec.containers.resources.requests`
+    * `unit` - the unit type of the resource type as inferred from the resource name in `spec.containers.limits`, such as `bytes`, `cores`. Empty if this is a unitless resource and no such calculation can be made.
+
+Specific resource implications:
+
+* `cpu` would be reported as the rate of usage over a reasonable window based on the scrape period of the kubelet - no window is perfect, but the summation is valuable to allow requests to be compared like to like.
+* `memory` would be reported as the working set bytes since that is the most accurate assessment of the pods usage and will in the majority of scenarios capture what the OOM killer will target. There may be some confusion because this metric may temporarily exceed total limit, but RSS often significantly underestimates the real memory workload usage.
+
+We use the generic metric form instead of a specific name containing the resource so that administrators can easily discover resources under the same metric.
+
+For alpha we would target cpu and memory to gain familiarity with this approach.
+
+### Test Plan
+
+The metrics implementation for lifecycle will be tested with unit tests to verify the correct transformation of pod -> metric for the relevant metrics. An integration test will verify that the active scheduler exposes the metrics on clusters where the scheduler metrics port is reachable. An e2e may be added later for clusters that expose the scheduler HTTP endpoint (which is both optional and may not be available in some configurations).
+
+### Graduation Criteria
+
+The metrics exposed at the new endpoint will be listed as alpha stability in 1.20, and may be changed or removed in subsequent releases. Default installations of Kubernetes should not utilize these metrics except for testing and feedback as they may change without notice
+
+#### Alpha -> Beta Graduation
```

Comment:
we also need to make sure that the memory overhead is reasonable, do you have estimates on that now?

Reply:
This only allocates when scraped (and implementation wise is reasonable within those limit) and does not hold extra memory.  So it has a minor and temporary memory impact.

Reply:
we need scalability tests to make sure that this doesn't crash the scheduler when scrapped.

Reply:
Algorithmically we have lots of O(pod) things we don’t scalability test.  I’m not opposed to some testing, but adding directed metrics translator is mostly a well understood problem domain for us.  Kube state metrics already provides orders of magnitude more metrics for less CPU, so at least at first pass for alpha the impact is minimal.